### ISSUE TITLE
Support for Expression pedal?

### ISSUE THREAD

**Issue by wigglepixel** [tlipinski/bitwig-novation-impulse#19]
Hi,

First of all thanks a lot for creating this script. It's a real life saver, because I need the sustain pedal to work and it doesn't work in the current impulse script coming with Bitwig. 

That said, I'd like to use the Expression pedal, connected to the Impulse, too. When testing with midimonitor.com the PC which has the Impulse connected is indeed receiving CC11 from changing the expression pedal which is connected to the Impulse61 keyboard. However, when trying to use that CC11 in the Pianoteq VST it doesn't receive it. 

So it looks like the bitwig-novation-impulse script doesn't forward that cc to Bitwig. Or I am missing something perhaps.

Am I missing something/doing something wrong, or is this not supported by the script yet? If the latter is the case, would it be possible to add support for this so we could use the second pedal input (the expression) of the keyboard too? It's pretty important for piano work and also super useful to use for other applications, like synths.

Thanks for considering!

Maarten

**Comment by tlipinski** [comment]
Hello @wigglepixel! Can you please tell what midi messages did you exactly see in midimonitor for expression pedal? Did you see some messages like `B20Axx` or something similar when you used the pedal? 
You can also take a look at script's console output in Bitwig and see whether messages are handled or not.
Here https://github.com/tlipinski/bitwig-novation-impulse/issues/17 I got a request for adding support for sustain pedal. With expression pedal it should be basically the same change to do.

**Comment by wigglepixel** [comment]
@tlipinski  Thanks for the quick reply. 

Looking in the midimonitor while pressing the expression Pedal I see:
![image](https://github.com/user-attachments/assets/b63f1858-bbae-4970-8cdc-a754036bb3e1)

Looking in the Bitwig Controller Script Console under 'Novation Impulse' there's nothing coming when pressing and releasing the expression pedal (not when pressing the sustain pedal either btw, while that one works just fine with your script)
![image](https://github.com/user-attachments/assets/a7e4257f-1545-45e5-9f41-c769dcfd1c4a)

Hope this helps. If you want more info just let me know.

Thanks again

**Comment by wigglepixel** [comment]
Ah, I see now that in the controller-setup of your addon I needed to enable debugging. 
After turning on debugging it's indeed logging 'unhandled' now when pressing the expression pedal.
![image](https://github.com/user-attachments/assets/d35441fd-0abb-4b1d-afeb-a69c9c18765f)

**Comment by tlipinski** [comment]
@wigglepixel that's exactly what I was looking for, thanks:) I've made test release and I wish you tested it https://github.com/tlipinski/bitwig-novation-impulse/releases/tag/v1.1-beta-expr. Instead of listing every single CC message type I simply allowed all of them to be handled by the script.
Please me know if test release makes expression pedal work and sustain pedal or mod wheel didn't stop working.

**Comment by wigglepixel** [comment]
@tlipinski Yes, with the new beta version the Expression pedal is working great. The sustain pedal is still working and I've also tested it together with the mod wheel and the pitch bend and all seems to work fine and together. I think it's a good decision to allow all CC's to be handled by the script. Sounds pretty flexible to me. 

Is this a version I can keep using, or are there special debug things or quick'n'dirty things insides making it less performant making it better for me to wait for a released version?

Either way, thanks a lot for the super quick action. Your script is awesome. Also like the 'send template'-feature to actually create a Bitwig template on the keyboard. Glad that the expression pedal is working now because I'd ordered a second pedal to use as expression pedal so it's nice to able to use it in Bitwig now. Keep it up and have a nice weekend!

**Comment by tlipinski** [comment]
Awesome, thank you for testing! :slightly_smiling_face:  You can keep using this version - it only has the needed change. I'll just rename it or create a new one with exactly the same code.
Have fun playing! :slightly_smiling_face:

**Comment by wigglepixel** [comment]
@tlipinski  Unfortunately, I just found an issue in the newer version, which is quite important: the transport controls aren't working anymore. When switching back to version 1.0 they are working again. And when switching back to version 1.1 they are broken again. So something must've been changed in the newer version which broke the transport controls.

BTW, not sure if this was needed, but when it didn't work, as a test, I did try it after resending the template, but without luck. Same issue remains.

**Comment by tlipinski** [comment]
Ok so probably I need to list expression pedal message explicitly. Please try https://github.com/tlipinski/bitwig-novation-impulse/releases/tag/v1.1-test

**Comment by wigglepixel** [comment]
@tlipinski  I've just tried the test-version; this is what I found for the time I had. I hope this helps. 

- Transport controls work again (and sustain pedal still works), but at a cost:

- expression pedal does not work

- rotator knobs, set to 'MIDI' don't send midi CC's --> 'unhandled'.
When set to 'Plugin' I don't see them working either?
(rotators here are important as they are used frequently by all kinds of parameters for synths and effects mainly)

- faders, when set to Mixer work fine for volumes, but when set to MIDI they don't send CC's --> 'unhandled'.
(Faders are used a lot for this, especially while playing strings to make it sound more realistic by controlling three CC's at the same time influencing the strings while playing. That's a well used workflow used some orchestral players, especialy when making music for movie-related work, like strings)

- solo and mute buttons also tested; all work, but: solo button doesn't respect setting Behaviour --> Solo --> Exclusive solo (multiple solos can be set, while the 'Exclusive solo' is enabled)


The rotator knobs and CC's coming from the faders are pretty important too. I didn't have much time to check if these worked in v1.0, but it looks like they didn't because you specifically only let through certain things? 
Honestly I think the workflow of letting every CC through is the best as there's just too many CC's we can control with the keyboard and we'd like to use the full potential of the keyboad. Many CC's are also quite needed.

I'm starting to get the feeling that this issue is starting to get related to this one:
https://github.com/tlipinski/bitwig-novation-impulse/issues/7
I didn't read the full thread there, but I think, if possible, it would be best if the fixed controllers would never be in the way of assigning CC's to other knobs and faders on the keyboard. Especially because many CC's have a fixed purpose by the MIDI standard, like sustain pedals.

It would be fine with me if these fixed controls (like the transport controls) are on CC-numbers that are set to 'undefined' by the MIDI standard though (https://anotherproducer.com/online-tools-for-musicians/midi-cc-list/). Looks like some of these are now in conflict with the MIDI standard, like the REC-button, which now sends CC32, which starnds for the LSB Controller according to the list on the link above. And for instance the third Mute or solo button which now sends CC11, which is reserverd for Expression pedals and so in conflict with the pedal. There might be more buttons/knobs in conflict with the standard now.

So in short: I think it would be best if users would never even have to worry about what CC's to assign to buttons and knobs, so that it never conflicts the fixed buttons like the transport controls. So for instance sending the fixed things via another midi channel might perhaps work? (Not sure if that's possible and if it's still 'plug and play' than). But if not possible at least the used CC's for the fixed buttons/knobs/faders I think should be on undefined CC's by the MIDI standard. That would prevent a lot of issues right away I think. And might also be the reason why things stopped working?

Please let me know if you need more info. I don't have much time left today, but can look at it later if you need more info.

**Comment by tlipinski** [comment]
Ok so first of all I completely messed up setting explicit midi code for expression pedal because I did it for CC10 instead of CC11 :wink: Such mistake could have broken something obviously.
Did rotator knobs and faders broke after my recent changes or did they work like that in previous version as well?
Please take another look at https://github.com/tlipinski/bitwig-novation-impulse/releases/tag/v1.1-test
I renamed previous version by adding `old1`. Use newest `NovationImpulse.bwextension`.

As you can see making changes without having gear anymore makes things really difficult even for simple stuff :wink: 
I hope I can at least make expression pedal finally working but as for the midi conflicts the chance is close to zero :disappointed:

**Comment by wigglepixel** [comment]
@tlipinski Sorry for the late response. I did some testing after your message and even recorded a video of me pressing/rotating/fading all physical buttons etc. on the machine, together with a screen recording of the bitwig console and a website showing the CC's in runtime. However, after doing so I only found out in the edit that the console was not set to your addon. So that was a bummer, because that wouldn't tell much and I didn't have more time to record another one and already took quite some time.

I'm currently not behind the system with the impulse, but I can tell you that the expression pedal in the latest test version was working again. 

Reason I didn't respond yet and wanted to create a video in the hopes of helping you, is because I honestly need all the faders and encoder knobs to transmit CC messages to control VST's/parameters (and let VST's treat them as pedals etc.) for reasons explained above. So if these aren't working the keyboard/controller basically isn't function as supposed to.

Unfortunately there don't seem to be other scripts around which have a complete implementation of the controller. One has a, another has b and c, but not d etc. Some have bugs, even after 9 years. 
And the whole reason I got here in the first place was because Bitwig itself delivers a script with the product wich is incomplete and full of bugs which even after 10 years (!!) aren't fixed. I've sent e-mails to Bitwig last weekend to (finally) fix this. And also pointed them towards this script which might help them to do so. But I am wondering if they even take fixing bugs seriously at this point, knowing that these bugs aren't fix in a decade while there are plenty of people facing these for years now (just a search on internet shows them).

So in short I now have a problem, because I can't use my controller as supposed to. So best scenario would be if either Bitwig could fix their or you could complete your script. But about your part I completely understand that that's difficult for you without having the device. And I already super appreciate that knowing that you even took the time to do this for me. 

So guess in the meantime, until Bitwig hopefully fixes the bugs, I guess my best bet is to use your script and need to accept that I can't use the full controller, or I need to switch back to Ableton Live 10 again (which is clearly not what I like and a huge step backwards).

Sorry for the long text, but thanks for all the great help. That's super appreciated and I fully understand your perspective. The pedals now both are working great, so maybe someday somebody could add the CC's to them or use this to make the script complete in Bitwig so we all can enjoy a nice working Impulse with Bitwig. Have a good day!

**Comment by tlipinski** [comment]
@wigglepixel I feel the pressure that I might be the only person to make it work;) I doubt that Bitwig is going to anything - they also probably no longer own any Impulse:D
Maybe let's just give it a try!
You had a good idea recording all midi messages from the console. I believe this way we could find all conflicts. Please share it when you record it:)

**Comment by wigglepixel** [comment]
@tlipinski Thanks, that's super apprecatied. Very kind of you. In that case I'll create the video with a cam on my keyboard where I push, fade and rotate all buttons/knobs/faders on the controller once and add a screenrecording of both the bitwig console as well as the visualisation of all midi data that is sent by the device. Of course after that I'll help you with testing and providing info you need. Would that be enough for you for now, or do you have any things you'd like to see or know besides this?

(BTW the Impulse keyboard is still sold new directly from Novation and also by resellers all over the world and used by many musicians for many years so I honestly see no reason for Bitwig not to have the controller there honestly, especially knowing that they deliver a script for it with the product)

Thanks again!

**Comment by wigglepixel** [comment]
@tlipinski  This one to let you know I've got 98% working on the impulse-bitwig now by a tip from the Bitwig Discord channel; 

There seems to be an extension out there making it possible to manually configure MIDI controllers for Bitwig; it's called 'DrivenByMoss / Flexi' (https://www.mossgrabers.de/Software/Bitwig/Bitwig.html). With this I managed to setup all the transport, fader- and rotary-controls with Bitwig. And it also forwards most basic features like keys/pads, pedals, mod wheel etc.

The thing that I couldn't get working is to have the feedback on the Impulse device, like having LEDs of the solo-buttons update when changed in Bitwig. There is a feature for feedback, but can't get it to work. However, I just set the solo-buttons to min 127 and max 127 so LEDs are always on, so never out of sync. Technically it just works. So for me this is enough to use everything on the keyboard.

BTW I've just purchased Studio One as well and in there there's no Impulse controller and no Flexi available. But it turns out that Novation created a firmware update where the Impulse keyboard can mimic Mackie HUI, making it possible to make the keyboard not only directly send all faders, knobs and transport controls, but even the fader-modes and knob-modes can be changed from their DAW-feature to CC codes and back (that's not possible on manually created keyboard templates).
Novation even created a new virtual MIDI in/output for this HUI part of the keyboard. The other part, so the keys, pads, pitch bender, pedals etc., can still be handled by a simple MIDI keyboard interface in Studio One. Making these two together handle the full keyboard. 

Knowing that Bitwig also supports the Mackie HUI and Simple Keyboard controllers Bitwig could do this mix-n-match too. So that way also in Bitwig I could probably make the fader-modes and knob-modes buttons working, as well as the up-down buttons to scroll through tracks etc. 
However, although I'll probably do that later, for now my setup with the Flexi-controller works just fine for what I need (and it already took me way too much time).

So I don't need this extension no more and for me there's no need no more to fix this. So thanks a lot for all the great work and willing to really help. Eventhough we don't even know each other, there was really no benefit for you and you don't even own the keyboard no more. That's very inspiring and appreciated. Keep up being you!

And for the rest; All the best for 2025!

**Comment by tlipinski** [comment]
@wigglepixel that's a great news that you finally managed to solve that issue this way or another! That was quite a long road:) Too bad that you're never exactly sure how your gear will be supported before buing it:/ Luckily it ended well in this case.

Thank you for your kind words and happy 2025!:)
